File: bicycle/wheel.py
class Wheel:
    BURST_PRESSURE : float = 100.0   # pounds per square inch
    DEFAULT_DIAMETER : float = 29.0  # inches
    DEFAULT_PRESSURE : float = 0.0   # pounds per square inch

    def __init__(self, diameter : float = DEFAULT_DIAMETER, pressure : float = DEFAULT_PRESSURE):
        self._diameter = diameter
        self._pressure = pressure
        self._burst : bool = False
        
    @property
    def diameter(self) -> float:
        return self._diameter
    
    @property
    def pressure(self) -> float:
        return self._pressure

    @pressure.setter
    def pressure(self,value : float) -> None:
        if value < 0:
            raise ValueError(f"pressure {value} must be non-negative")
        if value >= self.BURST_PRESSURE:
            self._burst = True
        if self._burst:
            value = 0.0
        self._pressure = value

    @property
    def flat(self) -> bool:
        return self._pressure == 0.0

    @property
    def burst(self) -> bool:
        return self._burst

File: cpp-diamond/src/Wheel.cpp


#include <stdexcept>
#include "Wheel.h"
namespace go
{

const double Wheel::BURST_PRESSURE = 100.0;
const double Wheel::DEFAULT_DIAMETER = 29.0;
const double Wheel::DEFAULT_PRESSURE = 0.0;

Wheel::Wheel(double _diameter, double _pressure)
    : diameter(_diameter), pressure(_pressure), burst(_pressure >= BURST_PRESSURE)
{
    if (pressure < 0)
    {
        throw std::invalid_argument("pressure cannot be negative");
    }
}

double Wheel::getDiameter() const
{
    return diameter;
}

void Wheel::setPressure(double value)
{
    if (value < 0.0)
    {
        throw std::invalid_argument("pressure cannot be negative");
    }
    if (value >= BURST_PRESSURE)
    {
        burst = true;
    }
    if (burst)
    {
        value = 0.0;
    }
    pressure = value;
}

double Wheel::getPressure() const
{
    return pressure;
}

bool Wheel::isFlat() const
{
    return pressure == 0.0;
}

bool Wheel::hasBurst() const
{
    return burst;
}

void Wheel::inflate(double additional)
{
    setPressure(getPressure() + additional);
}
} // namespace go
